### PR TITLE
feat: filter the Requests log by model and split out cancelled

### PR DESCRIPTION
**Pull request by guillaumegay13**
### ✨ What changed
- Model filter on the Requests log, multi-select, options from `filter-options`.
- `Cancelled` is its own status instead of being counted as `Failed`.
- Removed the Min/Max \$ cost inputs from the filter bar.

### 💭 Why
You could not ask the log which model a request ran on. Cost filtering by an absolute threshold needed you to already know the distribution you were trying to find, so it went.

### 👤 For users
Pick one or more models to narrow the log. The filter matches any provider attempt on the request, so filtering by the primary of a fallback chain still finds the request it was recovered on. A request Manifest blocked before any provider call matches on its requested model, which is what the Model column shows for those rows.

`status=failed` now returns fewer requests: cancelled ones moved to their own filter. They were previously swept in, because the request query read `failed` as "anything that is not a success". It now uses `sqlIsFailedStatus`, the same line the error taxonomy already drew.

### 🔧 For operators
No migration. `cost_min` and `cost_max` stay in the API and query service, so CLI, MCP and API callers keep them, only the dashboard inputs are gone. MCP `list_requests` picks up `cancelled` from the shared status enum.

### 📝 Notes
The distinct-model scan that already ran to derive the provider list is reused for the model options, so the dropdown costs no extra query. Models named only by attempt-less requests are folded in separately, cached on the same 5 minute TTL.
Column sorting is not here. It needs cost and token totals denormalised onto `requests`, which wants its own design pass.

---
## Summary by cubic
Adds a multi-select model filter to the Requests log and gives cancelled requests their own status instead of counting them as failed.

**Model filter**
- Model options come from `filter-options`, reusing the existing distinct-model scan and adding models that only appear on requests with no provider attempt.
- A request matches any provider attempt's model, or `requested_model` when it never reached a provider — the value the Model column shows for those rows.
- The `model` query parameter is comma-separated, trimmed, and capped at 50 models.
- Legacy installs without a requests table filter per attempt row, and the count cache is keyed by the model list so totals never mix across filters.
- The multi-select dropdown stays open after a pick and keeps picked models listed even when they leave the filtered window.

**Status and cost changes**
- `status=failed` now uses `sqlIsFailedStatus`, excluding cancelled and pending, and `status=cancelled` is a valid API/MCP filter value.
- The Min/Max $ inputs are removed from the filter bar; `cost_min` and `cost_max` stay in the API and query service for CLI, MCP, and API callers.
- No migration required.

<sup>Written for commit 9d0cf4b130eb248a0a47009d378c4794660de398. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/mnfst/llm-gateway/pull/2913?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: packages/backend/src/analytics/controllers/messages.controller.ts [hunk 1]
```diff
@@ -32,13 +42,8 @@ export class MessagesController {
...
       provider: query.provider,
-      connections: query.connections
-        ? query.connections
-            .split(',')
-            .map((id) => id.trim())
-            .filter(Boolean)
-            .slice(0, 50)
-        : undefined,
+      connections: splitCsv(query.connections),
+      models: splitCsv(query.model),
```
[the remark this answers is not in the document]
Comment (guillaumegay13) [reply packages/backend/src/analytics/controllers/messages.controller.ts:46]: Not reproducible: the spec passes (14/14 locally, and the Backend CI job is green on this commit). Jest's toHaveBeenCalledWith uses recursive equality that ignores undefined-valued properties, so `models: undefined` does not break the two exact call-shape expectations.

File: packages/frontend/src/pages/MessageLog.tsx [hunk 2]
```diff
@@ -471,6 +464,15 @@ const MessageLog: Component = () => {
     },
   );
 
+  // Models the tenant has actually used in range, labelled the way the Model
+  // column renders them. A model already picked stays listed even if it drops
+  // out of the window, so an active filter never loses its own option.
+  const modelOptions = createMemo<MultiSelectOption[]>(() => {
+    const available = messageFilterOptions()?.models ?? [];
+    const merged = [...new Set([...available, ...modelsFilter()])].sort();
+    return merged.map((model) => ({ value: model, label: getModelDisplayName(model) }));
```
[the remark this answers is not in the document]
Comment (guillaumegay13) [reply packages/frontend/src/pages/MessageLog.tsx:484]: Real, but not introduced here and deliberately consistent: getModelDisplayName reads a non-reactive cache everywhere it is used, including message-table-cells.tsx:249, which renders the Model column this filter pairs with, and ModelPrices.tsx:95. Making the filter reactive on its own would have it disagree with the column beside it. Worth fixing in model-display.ts for all callers at once, not here.

File: packages/frontend/src/pages/MessageLog.tsx [hunk 3]
```diff
@@ -271,6 +275,14 @@ const MessageLog: Component = () => {
     setAttemptStatusFilterValue(next);
     setSearchParams({ attempts: next || undefined }, { replace: true });
   };
+  // `?model=` narrows the log to requests that touched one of these models.
+  const [modelsFilter, setModelsFilterValue] = createSignal<string[]>(
+    typeof searchParams.model === 'string' ? searchParams.model.split(',').filter(Boolean) : [],
+  );
```
[the remark this answers is not in the document]
Comment (guillaumegay13) [reply packages/frontend/src/pages/MessageLog.tsx:281]: Fixed in 4294dbcf: modelsFilter now follows searchParams.model through a deferred effect, the same way status and range do. Covered by a new test that drives the param through a signal and asserts the next query carries the new model.

File: packages/frontend/tests/pages/MessageLog.test.tsx [hunk 4]
```diff
@@ -574,12 +533,64 @@ describe('MessageLog', () => {
...
+  });
+
+  it('offers Cancelled as its own status, distinct from Failed', async () => {
+    mockGetMessages.mockResolvedValue(messagesData);
+    const { container } = render(() => <MessageLog />);
+    await vi.waitFor(() => {
+      expect(container.textContent).toContain('msg-1234');
+    });
+    const status = selectWithOption(container, 'All statuses');
+    expect(status.textContent).toContain('Cancelled');
```
[the remark this answers is not in the document]
Comment (guillaumegay13) [reply packages/frontend/tests/pages/MessageLog.test.tsx:559]: Fixed in 4294dbcf: the test now selects Cancelled and asserts the API receives status=cancelled plus the URL sync, then selects Failed and asserts it still maps as before. The label assertion alone did not earn the test name.

File: packages/backend/src/analytics/services/messages-query.service.cost-filter.spec.ts [hunk 5]
```diff
@@ -93,6 +93,65 @@ describe('MessagesQueryService — cost filter edge cases', () => {
     return calls.find(([clause]) => typeof clause === 'string' && clause.includes(op));
   };
 
+  it('does not serve one model filter a total cached under another', async () => {
+    // getMessageCount() sends cache_total=true, so a count key that ignores the
```
[the remark this answers is not in the document]
Comment (guillaumegay13) [reply packages/backend/src/analytics/services/messages-query.service.cost-filter.spec.ts]: Fixed in 4294dbcf: reworded. getMessageCount lives in the frontend API client, so naming it from a backend spec sent readers looking for a symbol that is not in this package. The comment now describes the incoming request instead.

File: packages/backend/test/messages-model-filter.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+const ids = (body: { items: { id: string }[] }): string[] => body.items.map((i) => i.id);
+
+describe('GET /api/v1/messages?model=', () => {
+  it('returns only requests holding an attempt on that model', async () => {
+    const res = await request(app.getHttpServer())
+      .get(`/api/v1/messages?range=24h&model=${BETA}`)
+      .set('x-api-key', TEST_API_KEY)
+      .expect(200);
+
+    expect(ids(res.body)).toEqual(expect.arrayContaining([betaRequestId, fallbackRequestId]));
```
[the remark this answers is not in the document]
Comment (guillaumegay13) [reply packages/backend/test/messages-model-filter.e2e-spec.ts]: Fixed in 4294dbcf: asserts the exact set now, including reroutedRequestId, which does belong in the result since its only attempt ran on BETA. That is what pins the any-attempt semantics the test name claims.

File: packages/backend/src/analytics/services/messages-query.filter-options.spec.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    const options = await service.getMessageFilterOptions({ tenantId: 'tenant-1', range: '24h' });
+
+    expect(options.models).toEqual(['gpt-4o']);
+  });
+
+  it('scopes the blocked-model lookup to one harness when the caller names it', async () => {
+    const requestQb = makeRequestQb([{ model: 'gpt-4o' }]);
+    const { service } = makeService({ attemptModels: [], requestQb });
+
+    await service.getMessageFilterOptions({
```
[the remark this answers is not in the document]
Comment (guillaumegay13) [reply packages/backend/src/analytics/services/messages-query.filter-options.spec.ts:99]: Fixed in 4294dbcf: the harness-scoped test now also asserts the cutoff came from the caller range rather than the 90-day default, so a regression ignoring params.range fails here.

File: packages/backend/src/analytics/services/messages-query.filter-options.spec.ts [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+      agent_name: 'bot-1',
+    });
+
+    const clauses = requestQb.andWhere.mock.calls.map((call) => String(call[0]));
+    expect(clauses.some((clause) => clause.includes('r.agent_id = ('))).toBe(true);
+    expect(clauses.some((clause) => clause.includes('r.tenant_id = :blockedTenantId'))).toBe(true);
+    // The window comes from the caller's range, not the 90-day default: a
+    // regression that ignored params.range would otherwise pass every test here.
+    const cutoff = new Date(requestQb.where.mock.calls[0][1].cutoff as string).getTime();
+    expect(Date.now() - cutoff).toBeLessThan(25 * 60 * 60 * 1000);
```
[the remark this answers is not in the document]
Comment (guillaumegay13) [reply packages/backend/src/analytics/services/messages-query.filter-options.spec.ts]: Fixed in ccd80984. Confirmed the mechanism: computeCutoff emits local wall-clock with no zone (postgres-sql.ts:120), so re-parsing during a fall-back resolves the repeated hour to the earlier instant and the span reads 25h. The window is the repeated hour itself rather than the following 24h, but the assertion could still fail. Replaced the absolute clock difference with a comparison against computeCutoff of the 90-day default, which is exact, says what the test means, and cannot drift with DST.

File: packages/frontend/src/components/MultiSelect.tsx [hunk 9]
```diff
@@ -45,7 +45,17 @@ const MultiSelect: Component<MultiSelectProps> = (props) => {
...
+    // This runs after the option's own handler. When a caller derives its
+    // options from the selection — the Requests log's model filter does, so a
+    // picked model stays listed even if it leaves the range window — toggling
+    // rebuilds the option nodes and the clicked one is already detached by now.
+    // `contains` would call that "outside" and shut the menu after one pick, so
+    // a node that left the document mid-click is never treated as an outside
+    // click. Stopping propagation on the dropdown cannot do this job: Solid
+    // delegates onClick to its own document listener, not the element.
+    if (!target.isConnected) return;
+    if (ref && !ref.contains(target)) setOpen(false);
```
[the remark this answers is not in the document]
Comment (guillaumegay13) [reply packages/frontend/src/components/MultiSelect.tsx]: Right, and reproduced before fixing: an outside button that removes itself on click left the menu open under the isConnected guard. Switched to composedPath(), which is the dispatch-time snapshot, so a replaced option still resolves as inside while a self-removing outside target still closes the menu. Falls back to contains() if the path is empty. Added a test for the self-removing case alongside the existing rebuild and plain-outside-click ones. Fixed in 9d0cf4b1.
